> **Oracle JDK 11 and higher**
> "Keycloak Migration TestSuite For Server Config" is failing with the following compilation error.
> 
> ```
> [ERROR] Failed to execute goal org.wildfly.plugins:wildfly-maven-plugin:1.1.0.Final:start (start-stop-standalone) on project server-config-migration: Execution start-stop-standalone of goal org.wildfly.plugins:wildfly-maven-plugin:1.1.0.Final:start failed: Plugin org.wildfly.plugins:wildfly-maven-plugin:1.1.0.Final or one of its dependencies could not be resolved: Could not find artifact sun.jdk:jconsole:jar:jdk at specified path /Library/Java/JavaVirtualMachines/jdk-11.0.4.jdk/Contents/Home/../lib/jconsole.jar -> [Help 1]
> ```
> 
> This is however "just" a test module which is even disabled by default. Even though it's not ideal, we could fix this in a follow-up task.

@vmuzikar thanks for reviewing this.

@hmlnarik @mposolda considering that this module is disabled by default. Do you guys think it is fine to follow up in another Jira, so we can move on with this change?

> 
> **Oracle JDK 12 and higher**
> "Keycloak Spring Security Integration" is failing with the following compilation error.

@vmuzikar despite the Jira description, is there any other reason to test this PR with JDK 12 or higher? The reason why I'm asking is because I would like us to focus only on JDK 11 get it done and later think about other versions if needed.

> 
> ```
> [INFO] Compiling 38 source files to /Users/vmuzikar/Documents/keycloak/adapters/oidc/spring-security/target/classes
> [INFO] -------------------------------------------------------------
> [WARNING] COMPILATION WARNING : 
> [INFO] -------------------------------------------------------------
> [WARNING] bootstrap class path not set in conjunction with -source 6
> [INFO] 1 warning
> [INFO] -------------------------------------------------------------
> [INFO] -------------------------------------------------------------
> [ERROR] COMPILATION ERROR : 
> [INFO] -------------------------------------------------------------
> [ERROR] Source option 6 is no longer supported. Use 7 or later.
> [ERROR] Target option 6 is no longer supported. Use 7 or later.
> [INFO] 2 errors 
> ```
> 
> I haven't discovered this sooner because before that it never got to build this module as other errors appeared much sooner in other models (which skipped this module).
> 
> @douglaspalmer Please let me know if you need any help with verifying or reproducing this on your local machine. Perhaps we could schedule a brief call to resolve it more quickly.

--- a/adapters/oidc/pom.xml
+++ b/adapters/oidc/pom.xml
@@ -32,7 +32,6 @@
 
     <modules>
         <module>adapter-core</module>
-        <module>as7-eap6</module>
         <module>installed</module>
         <module>fuse7</module>
         <module>kcinit</module>
@@ -52,4 +51,16 @@
         <module>wildfly</module>
         <module>wildfly-elytron</module>
     </modules>
+
+    <profiles>
+        <profile>
+            <id>AS7</id>
+            <activation>
+                <jdk>[,9)</jdk>
+            </activation>
+            <modules>
+                <module>as7-eap6</module>
+            </modules>
+        </profile>
+    </profiles>
 </project>

--- a/adapters/oidc/spring-security/pom.xml
+++ b/adapters/oidc/spring-security/pom.xml
@@ -128,17 +128,4 @@
             <scope>test</scope>
         </dependency>
     </dependencies>
-    <build>
-        <plugins>
-            <plugin>
-                <groupId>org.apache.maven.plugins</groupId>
-                <artifactId>maven-compiler-plugin</artifactId>
-                <configuration>
-                    <source>1.6</source>
-                    <target>1.6</target>
-                </configuration>
-            </plugin>
-        </plugins>
-    </build>
-
 </project>

--- a/adapters/saml/core/src/test/java/org/keycloak/adapters/cloned/HttpAdapterUtilsTest.java
+++ b/adapters/saml/core/src/test/java/org/keycloak/adapters/cloned/HttpAdapterUtilsTest.java
@@ -8,9 +8,11 @@ package org.keycloak.adapters.cloned;
 import java.io.InputStream;
 import java.security.cert.X509Certificate;
 import java.util.List;
+import javax.xml.crypto.XMLStructure;
 import javax.xml.crypto.dsig.keyinfo.KeyInfo;
 import javax.xml.crypto.dsig.keyinfo.KeyName;
 import javax.xml.crypto.dsig.keyinfo.X509Data;
+import org.hamcrest.Matcher;
 import static org.hamcrest.CoreMatchers.*;
 import org.junit.Test;
 import static org.junit.Assert.*;
@@ -25,7 +27,7 @@ import org.keycloak.saml.common.exceptions.ParsingException;
  */
 public class HttpAdapterUtilsTest {
 
-    private <T> T getContent(List<Object> objects, Class<T> clazz) {
+    private <T> T getContent(List<?> objects, Class<T> clazz) {
         for (Object o : objects) {
             if (clazz.isInstance(o)) {
                 return (T) o;
@@ -49,10 +51,14 @@ public class HttpAdapterUtilsTest {
         X509Data x509data;
         X509Certificate x509certificate;
 
+        Matcher<Iterable<? super XMLStructure>> x509DataMatcher = hasItem(instanceOf(X509Data.class));
+        Matcher<Iterable<? super XMLStructure>> keyNameMatcher = hasItem(instanceOf(KeyName.class));
+
         ki = res.get(KeycloakSamlAdapterV1QNames.ATTR_SIGNING.getQName().getLocalPart()).get(0);
         assertThat(ki.getContent().size(), equalTo(2));
-        assertThat((List<Object>) ki.getContent(), hasItem(instanceOf(X509Data.class)));
-        assertThat((List<Object>) ki.getContent(), hasItem(instanceOf(KeyName.class)));
+
+        assertThat((Iterable<? super XMLStructure>) ki.getContent(), x509DataMatcher);
+        assertThat((Iterable<? super XMLStructure>) ki.getContent(), keyNameMatcher);
 
         keyName = getContent(ki.getContent(), KeyName.class);
         assertThat(keyName.getName(), equalTo("rJkJlvowmv1Id74GznieaAC5jU5QQp_ILzuG-GsweTI"));
@@ -65,8 +71,8 @@ public class HttpAdapterUtilsTest {
 
         ki = res.get(KeycloakSamlAdapterV1QNames.ATTR_SIGNING.getQName().getLocalPart()).get(1);
         assertThat(ki.getContent().size(), equalTo(2));
-        assertThat((List<Object>) ki.getContent(), hasItem(instanceOf(X509Data.class)));
-        assertThat((List<Object>) ki.getContent(), hasItem(instanceOf(KeyName.class)));
+        assertThat((Iterable<? super XMLStructure>) ki.getContent(), x509DataMatcher);
+        assertThat((Iterable<? super XMLStructure>) ki.getContent(), keyNameMatcher);
 
         keyName = getContent(ki.getContent(), KeyName.class);
         assertThat(keyName.getName(), equalTo("BzYc4GwL8HVrAhNyNdp-lTah2DvU9jU03kby9Ynohr4"));

--- a/adapters/saml/pom.xml
+++ b/adapters/saml/pom.xml
@@ -36,7 +36,6 @@
         <module>undertow</module>
         <module>tomcat</module>
         <module>wildfly</module>
-        <module>as7-eap6</module>
         <module>servlet-filter</module>
         <module>wildfly-elytron</module>
     </modules>
@@ -51,6 +50,16 @@
             </activation>
             <modules>
                 <module>jetty</module>
+            </modules>
+        </profile>
+
+        <profile>
+            <id>AS7</id>
+            <activation>
+                <jdk>[,9)</jdk>
+            </activation>
+            <modules>
+                <module>as7-eap6</module>
             </modules>
         </profile>
     </profiles>

--- a/common/pom.xml
+++ b/common/pom.xml
@@ -68,6 +68,10 @@
             <artifactId>hamcrest-all</artifactId>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>com.sun.activation</groupId>
+            <artifactId>javax.activation</artifactId>
+        </dependency>
     </dependencies>
     <build>
         <resources>

--- a/distribution/adapters/pom.xml
+++ b/distribution/adapters/pom.xml
@@ -31,7 +31,6 @@
     <packaging>pom</packaging>
 
     <modules>
-        <module>as7-eap6-adapter</module>
         <module>fuse-adapter-zip</module>
         <module>js-adapter-zip</module>
         <module>osgi</module>
@@ -55,5 +54,14 @@
                 <module>js-adapter-npm-zip</module>
             </modules>
         </profile>
+        <profile>
+             <id>AS7</id>
+             <activation>
+                 <jdk>(,9]</jdk>
+             </activation>
+             <modules>
+                 <module>as7-eap6-adapter</module>
+             </modules>
+         </profile>
     </profiles>
 </project>

--- a/distribution/saml-adapters/pom.xml
+++ b/distribution/saml-adapters/pom.xml
@@ -32,7 +32,6 @@
 
     <modules>
         <module>wildfly-adapter</module>
-        <module>as7-eap6-adapter</module>
     </modules>
 
     <profiles>
@@ -51,6 +50,15 @@
                 <module>tomcat7-adapter-zip</module>
             </modules>
         </profile>
+        <profile>
+             <id>AS7</id>
+             <activation>
+                 <jdk>(,9]</jdk>
+             </activation>
+             <modules>
+                 <module>as7-eap6-adapter</module>
+             </modules>
+         </profile>
     </profiles>
 
 </project>

--- a/pom.xml
+++ b/pom.xml
@@ -84,6 +84,7 @@
         <slf4j.version>1.7.22</slf4j.version>
         <sun.istack.version>3.0.7</sun.istack.version>
         <sun.jaxb.version>2.3.1</sun.jaxb.version>
+        <sun.activation.version>1.2.0</sun.activation.version>
         <org.glassfish.jaxb.xsom.version>2.3.1</org.glassfish.jaxb.xsom.version>
         <undertow.version>2.0.21.Final</undertow.version>
         <elytron.version>1.9.1.Final</elytron.version>
@@ -141,7 +142,7 @@
         <jmeter.analysis.plugin.version>1.0.4</jmeter.analysis.plugin.version>
         <minify.plugin.version>1.7.6</minify.plugin.version>
         <osgi.bundle.plugin.version>2.3.7</osgi.bundle.plugin.version>
-        <wildfly.plugin.version>1.1.0.Final</wildfly.plugin.version>
+        <wildfly.plugin.version>1.2.1.Final</wildfly.plugin.version>
         <nexus.staging.plugin.version>1.6.5</nexus.staging.plugin.version>
         <frontend.plugin.version>1.5</frontend.plugin.version>
         <docker.maven.plugin.version>0.28.0</docker.maven.plugin.version>
@@ -253,9 +254,30 @@
                 <version>${sun.istack.version}</version>
             </dependency>
             <dependency>
+                <groupId>javax.xml.bind</groupId>
+                <artifactId>jaxb-api</artifactId>
+                <version>${sun.jaxb.version}</version>
+            </dependency>
+            <dependency>
+                <groupId>com.sun.xml.ws</groupId>
+                <artifactId>rt</artifactId>
+                <version>${sun.jaxb.version}</version>
+                <exclusions>
+                    <exclusion>
+                        <groupId>com.fasterxml.woodstox</groupId>
+                        <artifactId>woodstox-core</artifactId>
+                    </exclusion>
+                </exclusions>
+            </dependency>
+            <dependency>
                 <groupId>com.sun.xml.bind.external</groupId>
                 <artifactId>rngom</artifactId>
                 <version>${sun.jaxb.version}</version>
+            </dependency>
+            <dependency>
+                <groupId>com.sun.activation</groupId>
+                <artifactId>javax.activation</artifactId>
+                <version>${sun.activation.version}</version>
             </dependency>
             <dependency>
                 <groupId>org.glassfish.jaxb</groupId>

--- a/saml-core-api/pom.xml
+++ b/saml-core-api/pom.xml
@@ -52,6 +52,14 @@
             <groupId>org.apache.santuario</groupId>
             <artifactId>xmlsec</artifactId>
         </dependency>
+        <dependency>
+            <groupId>javax.xml.bind</groupId>
+            <artifactId>jaxb-api</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>com.sun.xml.ws</groupId>
+            <artifactId>rt</artifactId>
+        </dependency>
     </dependencies>
     <build>
         <resources>

--- a/saml-core/pom.xml
+++ b/saml-core/pom.xml
@@ -73,6 +73,14 @@
             <version>2.6</version>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>javax.xml.bind</groupId>
+            <artifactId>jaxb-api</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>com.sun.xml.ws</groupId>
+            <artifactId>rt</artifactId>
+        </dependency>
     </dependencies>
     <build>
         <resources>

--- a/saml-core/src/main/java/org/keycloak/saml/processing/api/util/KeyInfoTools.java
+++ b/saml-core/src/main/java/org/keycloak/saml/processing/api/util/KeyInfoTools.java
@@ -34,7 +34,7 @@ public class KeyInfoTools {
      * @param clazz
      * @return The object or {@code null} if not found.
      */
-    public static <T> T getContent(Iterable<Object> objects, Class<T> clazz) {
+    public static <T> T getContent(Iterable<?> objects, Class<T> clazz) {
         if (objects == null) {
             return null;
         }

--- a/saml-core/src/main/java/org/keycloak/saml/processing/core/util/XMLSignatureUtil.java
+++ b/saml-core/src/main/java/org/keycloak/saml/processing/core/util/XMLSignatureUtil.java
@@ -41,6 +41,7 @@ import org.w3c.dom.NodeList;
 import org.xml.sax.SAXException;
 
 import javax.xml.bind.JAXBException;
+import javax.xml.crypto.XMLStructure;
 import javax.xml.crypto.MarshalException;
 import javax.xml.crypto.dsig.CanonicalizationMethod;
 import javax.xml.crypto.dsig.DigestMethod;
@@ -740,7 +741,7 @@ public class XMLSignatureUtil {
     private static KeyInfo createKeyInfo(String keyName, PublicKey publicKey, X509Certificate x509Certificate) throws KeyException {
         KeyInfoFactory keyInfoFactory = fac.getKeyInfoFactory();
 
-        List<Object> items = new LinkedList<>();
+        List<XMLStructure> items = new LinkedList<>();
 
         if (keyName != null) {
             items.add(keyInfoFactory.newKeyName(keyName));

--- a/testsuite/integration-arquillian/test-apps/test-apps-dist/pom.xml
+++ b/testsuite/integration-arquillian/test-apps/test-apps-dist/pom.xml
@@ -78,7 +78,6 @@
                         </configuration>
                     </execution>
                 </executions>
-                <version>2.5.5</version>
             </plugin>
         </plugins>
     </build>

--- a/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/adapter/servlet/SAMLServletAdapterTest.java
+++ b/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/adapter/servlet/SAMLServletAdapterTest.java
@@ -1703,8 +1703,8 @@ public class SAMLServletAdapterTest extends AbstractSAMLServletAdapterTest {
 
         //printDocument(authnRequestMessage.getSOAPPart().getContent(), System.out);
 
-        Iterator<SOAPHeaderElement> it = authnRequestMessage.getSOAPHeader().<SOAPHeaderElement>getChildElements(new QName("urn:oasis:names:tc:SAML:2.0:profiles:SSO:ecp", "Request"));
-        SOAPHeaderElement ecpRequestHeader = it.next();
+        Iterator<javax.xml.soap.Node> it = authnRequestMessage.getSOAPHeader().<SOAPHeaderElement>getChildElements(new QName("urn:oasis:names:tc:SAML:2.0:profiles:SSO:ecp", "Request"));
+        SOAPHeaderElement ecpRequestHeader = (SOAPHeaderElement)it.next();
         NodeList idpList = ecpRequestHeader.getElementsByTagNameNS("urn:oasis:names:tc:SAML:2.0:protocol", "IDPList");
 
         Assert.assertThat("No IDPList returned from Service Provider", idpList.getLength(), is(1));
@@ -1790,13 +1790,13 @@ public class SAMLServletAdapterTest extends AbstractSAMLServletAdapterTest {
                 .get();
 
         SOAPMessage authnRequestMessage = MessageFactory.newInstance().createMessage(null, new ByteArrayInputStream(authnRequestResponse.readEntity(byte[].class)));
-        Iterator<SOAPHeaderElement> it = authnRequestMessage.getSOAPHeader().<SOAPHeaderElement>getChildElements(new QName("urn:liberty:paos:2003-08", "Request"));
+        Iterator<javax.xml.soap.Node> it = authnRequestMessage.getSOAPHeader().<SOAPHeaderElement>getChildElements(new QName("urn:liberty:paos:2003-08", "Request"));
 
         it.next();
 
         it = authnRequestMessage.getSOAPHeader().<SOAPHeaderElement>getChildElements(new QName("urn:oasis:names:tc:SAML:2.0:profiles:SSO:ecp", "Request"));
-        SOAPHeaderElement ecpRequestHeader = it.next();
-        NodeList idpList = ecpRequestHeader.getElementsByTagNameNS("urn:oasis:names:tc:SAML:2.0:protocol", "IDPList");
+        javax.xml.soap.Node ecpRequestHeader = it.next();
+        NodeList idpList = ((SOAPHeaderElement)ecpRequestHeader).getElementsByTagNameNS("urn:oasis:names:tc:SAML:2.0:protocol", "IDPList");
 
         Assert.assertThat("No IDPList returned from Service Provider", idpList.getLength(), is(1));
 

